File: 2016/Day10/main.dart
import 'dart:collection';
import 'dart:io';

class Bot {
  int low = -1;
  int high = -1;
  int lowTarget = -1;
  int highTarget = -1;
  bool lowIsOutput = false;
  bool highIsOutput = false;
}

Map<int, Bot> bots = {};
Map<int, int> outputs = {};

void giveChip(int botId, int value) {
  Bot bot = bots.putIfAbsent(botId, () => Bot());
  if (bot.low == -1) {
    bot.low = value;
  } else {
    bot.high = value;
    if (bot.low > bot.high) {
      int temp = bot.low;
      bot.low = bot.high;
      bot.high = temp;
    }
  }
  bots[botId] = bot;
}

void processBot(int botId) {
  Bot bot = bots[botId]!;
  if (bot.low != -1 && bot.high != -1) {
    if (bot.low == 17 && bot.high == 61) {
      print(
        "Bot $botId is responsible for comparing value-61 microchips and value-17 microchips",
      );
    }

    if (bot.lowIsOutput) {
      outputs[bot.lowTarget] = bot.low;
    } else {
      giveChip(bot.lowTarget, bot.low);
      processBot(bot.lowTarget);
    }

    if (bot.highIsOutput) {
      outputs[bot.highTarget] = bot.high;
    } else {
      giveChip(bot.highTarget, bot.high);
      processBot(bot.highTarget);
    }

    bot.low = -1;
    bot.high = -1;
  }
  bots[botId] = bot;
}

void main() async {
  Queue<List<int>> initialChips = Queue();

  List<String> lines = await File('input.txt').readAsLines();

  for (String line in lines) {
    List<String> words = line.split(' ');

    if (words[0] == 'value') {
      int value = int.parse(words[1]);
      int botId = int.parse(words[5]);
      initialChips.add([botId, value]);
    } else if (words[0] == 'bot') {
      int botId = int.parse(words[1]);
      String lowType = words[5];
      int lowTarget = int.parse(words[6]);
      String highType = words[10];
      int highTarget = int.parse(words[11]);

      Bot bot = bots.putIfAbsent(botId, () => Bot());
      bot.lowTarget = lowTarget;
      bot.highTarget = highTarget;
      bot.lowIsOutput = (lowType == 'output');
      bot.highIsOutput = (highType == 'output');
      bots[botId] = bot;
    }
  }

  while (initialChips.isNotEmpty) {
    List<int> instruction = initialChips.removeFirst();
    int botId = instruction[0];
    int value = instruction[1];
    giveChip(botId, value);
    processBot(botId);
  }

  final product = outputs[0]! * outputs[1]! * outputs[2]!;
  print("The product of the values in outputs 0, 1, and 2 is: $product");
}

File: 2016/Day10/main.lua
local bots = {}
local outputs = {}

---@param bot_id number
---@return table
local function get_bot(bot_id)
    if not bots[bot_id] then
        bots[bot_id] = {
            low = -1,
            high = -1,
            low_target = -1,
            high_target = -1,
            low_is_output = false,
            high_is_output = false,
        }
    end
    return bots[bot_id]
end

---@param bot_id number
---@param value number
---@return nil
local function give_chip(bot_id, value)
    local bot = get_bot(bot_id)
    if (bot.low == -1) then
        bot.low = value
    else
        bot.high = value
        if bot.low > bot.high then
            bot.low, bot.high = bot.high, bot.low
        end
    end
    bots[bot_id] = bot
end

---@param bot_id number
---@return nil
local function process_bot(bot_id)
    local bot = bots[bot_id]
    if bot.low ~= -1 and bot.high ~= -1 then
        if bot.low == 17 and bot.high == 61 then
            print('Bot ' .. bot_id .. ' is responsible for comparing value-17 microchips and value-61 microchips')
        end

        if bot.low_is_output then
            outputs[bot.low_target] = bot.low
        else
            give_chip(bot.low_target, bot.low)
            process_bot(bot.low_target)
        end

        if bot.high_is_output then
            outputs[bot.high_target] = bot.high
        else
            give_chip(bot.high_target, bot.high)
            process_bot(bot.high_target)
        end

        bot.low = -1
        bot.high = -1
    end
    bots[bot_id] = bot
end

local initial_chips = {}

for line in io.lines('input.txt') do
    local words = {}
    for word in line:gmatch('%S+') do
        table.insert(words, word)
    end

    if words[1] == 'value' then
        local value = tonumber(words[2])
        local bot_id = tonumber(words[6])
        table.insert(initial_chips, { bot_id, value })
    elseif words[1] == 'bot' then
        local bot_id = tonumber(words[2])
        local low_type = words[6]
        local low_target = tonumber(words[7])
        local high_type = words[11]
        local high_target = tonumber(words[12])

        assert(bot_id, 'bot_id is nil')

        local bot = get_bot(bot_id)
        bot.low_target = low_target
        bot.high_target = high_target
        bot.low_is_output = (low_type == 'output')
        bot.high_is_output = (high_type == 'output')
        bots[bot_id] = bot
    end
end

for _, instruction in ipairs(initial_chips) do
    local bot_id, value = instruction[1], instruction[2]
    give_chip(bot_id, value)
    process_bot(bot_id)
end

print('The product of the values in outputs 0, 1, and 2 is: ' .. (outputs[0] * outputs[1] * outputs[2]))
